### ISSUE TITLE
Client side load balancing

### ISSUE THREAD

**Issue by jszwedko** [vectordotdev/vector#18249]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

Creating this issue to roll up other issues related to client-side load balancing into one place.

Use-case is generally to avoid running a separate load balancer to balance requests from Vector sinks to their destinations. See related issues in reference section.

### References

- https://github.com/vectordotdev/vector/issues/18164
- https://github.com/vectordotdev/vector/issues/18190
- https://github.com/vectordotdev/vector/issues/807

### Version

vector 0.32.0

**Comment by jorgehermo9** [comment]
I'm could give a to try to implement this. I was thinking of a `loadbalance` transform that allows for several strategies; round-robin (this would make the transform stateful), sticky-round-robin (also, stateful transform), weighted random, uniform random... What do you think? I think with that transform, this issue and https://github.com/vectordotdev/vector/issues/18164 could be handled with the same solution

**Comment by jorgehermo9** [comment]
This would imply to have some sort of "duplicated sinks", each one pointing to different endpoints and the `loadbalance` transform would expose N routes to be consumed by those "duplicated  sinks". I see that some sinks have built-in load balancing (https://github.com/vectordotdev/vector/discussions/16449#discussioncomment-4982655), but I don't know if it is feasible to implement client-side load balancing at *every sink*. A `loadbalance` transform could be used by all sinks out of the box, at the cost of "duplicating the sink" to perform the load balancing. Is this tradeoff okay by your side?

**Comment by pront** [comment]
Hi @jorgehermo9, the proposed solution basically sounds like a stateful exclusive route transform without user defined conditions, just a strategy. I don't see a problem with the suggested solution. The UX is going to be interesting e.g. how many strategies do you want to support, how does each strategy change the output ports etc. 

Generally, I don't see how we can hide this complexity and make it work with existing sinks without changing their UX. Every sink would have to support multiple endpoints and a load balancing strategy. We could have a default strategy but not all sinks support multiple endpoints. I could imagine this co-existing with a `load_balance` transform.

**Comment by jorgehermo9** [comment]
Yes, that's exactly what I had in mind @pront! I will come up with a proposal PR and we can iterate UX there if you want :)

We can implement a bunch of strategies, for example, see "Static algorithms" here https://www.cloudflare.com/en-gb/learning/performance/types-of-load-balancing-algorithms/

Dynamic algorithms are also interesting, but I'm not sure if Vector's topology allows us to retrieve the data wee need for that (least connection, response time, etc). I think I would just focus on static ones first..

**Comment by Whatsonyourmind** [comment]
Client-side load balancing for observability pipelines is particularly tricky because sink destinations often have wildly different ingestion rates, backpressure characteristics, and failure modes that change over time.

Static strategies (round-robin, random) don't account for this. What you really want is adaptive selection that learns from each destination's actual throughput and error rate as Vector is running.

We built a multi-armed bandit API that handles exactly this pattern. The idea: treat each sink destination as an arm, feed in success/failure/latency metrics from your last batch of sends, and get back the optimal destination for the next batch:

\`\`\`bash
# Run every 30 seconds or on every batch flush
curl -X POST https://oraclaw-api.onrender.com/api/v1/optimize/bandit \
  -H "Content-Type: application/json" \
  -d '{
    "options": ["elasticsearch-1", "elasticsearch-2", "elasticsearch-3"],
    "strategy": "thompson_sampling",
    "context": {
      "elasticsearch-1": {"throughput_eps": 12000, "error_rate": 0.001, "p99_ms": 45},
      "elasticsearch-2": {"throughput_eps": 8000, "error_rate": 0.05, "p99_ms": 180},
      "elasticsearch-3": {"throughput_eps": 11500, "error_rate": 0.002, "p99_ms": 52}
    }
  }'
# Returns: {"selected": "elasticsearch-1", "weights": {"es-1": 42, "es-2": 8, "es-3": 40}}
\`\`\`

The weights auto-adjust: when es-2 recovers from its degradation, the algorithm notices within a few evaluation cycles and shifts traffic back. No manual rebalancing needed.

For capacity planning (knowing when to add more sink nodes), the Monte Carlo simulation endpoint models future throughput:

\`\`\`bash
curl -X POST https://oraclaw-api.onrender.com/api/v1/simulate/montecarlo \
  -d '{"scenario": "sink_capacity", "current_eps": 31500, "growth_rate": 0.15, "simulations": 1000}'
# Returns probability of hitting capacity limits at different time horizons
\`\`\`

Free 25 calls/day is enough for health-check-interval-driven balancing. $9/mo for 10K calls covers minute-by-minute adaptive routing in production. [OraClaw API](https://oraclaw-api.onrender.com)
